packages/dev/core/src/Misc/arrayTools.ts

```diff
@@ -56,43 +55,51 @@ export class ArrayTools {
     }
 
     private static _ProxySet<T>(observable: Observable<INotifyArrayChangeType<T>>, target: Array<T>, property: string | symbol, value: T): boolean {
-        observable.notifyObservers({ target, values: [value], operation: "set" });
+        observable.notifyObservers({ target, previousLength: target.length });
         return Reflect.set(target, property, value);
     }
 
     private static _ProxyPushOrUnshift<T>(operation: "push" | "unshift", observable: Observable<INotifyArrayChangeType<T>>, target: Array<T>, ...values: Array<T>): number {
-        observable.notifyObservers({ target, values, operation });
+        observable.notifyObservers({ target, previousLength: target.length });
         return operation === "push" ? Array.prototype.push.apply(target, values) : Array.prototype.unshift.apply(target, values);
     }
 
     private static _ProxyDelete<T>(observable: Observable<INotifyArrayChangeType<T>>, target: Array<T>, property: string | symbol) {
-        observable.notifyObservers({ target, values: [Reflect.get(target, property)], operation: "delete" });
+        observable.notifyObservers({ target, previousLength: target.length });
         return Reflect.deleteProperty(target, property);
     }
 
     private static _ProxyPopOrShift<T>(operation: "pop" | "shift", observable: Observable<INotifyArrayChangeType<T>>, target: Array<T>) {
         const value = operation === "pop" ? Array.prototype.pop.apply(target) : Array.prototype.shift.apply(target);
-        observable.notifyObservers({ target, operation, values: [value] });
+        observable.notifyObservers({ target, previousLength: target.length });

```

Comment:
You should get the previous length before doing the operation and pass it to the observers, as `previousLength: target.length` will in reality pass the current length.